fe/fe-core/src/main/java/org/apache/doris/nereids/pattern/GroupMatching.java

```diff
@@ -0,0 +1,105 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package org.apache.doris.nereids.pattern;
+
+import org.apache.doris.nereids.memo.Group;
+import org.apache.doris.nereids.memo.GroupExpression;
+import org.apache.doris.nereids.operators.OperatorType;
+import org.apache.doris.nereids.trees.TreeNode;
+
+import com.google.common.collect.Lists;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
+/**
+ * Get all pattern matching subtree in query plan from a group.
+ */
+public class GroupMatching<NODE_TYPE extends TreeNode> implements Iterable<NODE_TYPE> {
+    private final Pattern pattern;
+    private final Group group;
+
+    public GroupMatching(Pattern pattern, Group group) {
+        this.pattern = Objects.requireNonNull(pattern);
+        this.group = Objects.requireNonNull(group);
+    }
+
+    public Iterator<NODE_TYPE> iterator() {
+        return new GroupIterator<>(pattern, group);
+    }
+
+    /**
+     * Iterator to get all subtrees from a group.
+     */
+    public static class GroupIterator<NODE_TYPE extends TreeNode> implements Iterator<NODE_TYPE> {
```

Comment:
Please add example in comment

Reply:
UTs and Java docs will added  when Nereids is stable